Rewrite and split the Overview article (#2834)
surveyjs/survey-creator: docs/property-grid.md

```js
const blackList = [ "visible", "isRequired" ];
// const whiteList = [ "title", "name" ];

creator.onShowingProperty.add(function (_, options) {
  if (options.obj.getType() == "panel") {
    // Hide properties found in `blackList`
    options.canShow = blackList.indexOf(options.property.name) < 0;

    // Hide all properties except those found in `whiteList`
    // options.canShow = whiteList.indexOf(options.property.name) > -1;
  }
});
```

<div id="addproperties"></div>

## Add Custom Properties to the Property Grid

Custom properties can be serialized and included in the survey JSON definition. To add a custom property, call the `addProperty(questionType, propertySettings)` method on the `Survey.Serializer` object. This method accepts the following arguments:

- `questionType`        

> A string value that specifies the question type to which the property should be added. You can use a specific type (see the [getType](https://surveyjs.io/Documentation/Library?id=Question#getType) description) or one of the base types. In the latter case, the new property will be added to all question types that derive from the base type. Refer to the API of a specific question type for information on its inheritance chain. For example, the following image illustrates the inheritance chain of the [Text](https://surveyjs.io/Documentation/Library?id=questiontextmodel) question type:

Reviewer: ```suggestion
A string value that specifies the question type to which the property should be added. You can use a specific type (see the [getType](https://surveyjs.io/Documentation/Library?id=Question#getType) description) or one of the base types. In the latter case, the new property is added to all question types derived from the base type. Refer to the API of a specific question type for information on its inheritance chain. For example, the following image illustrates the inheritance chain of the [Text](https://surveyjs.io/Documentation/Library?id=questiontextmodel) question type:
```

| `"expression"` | Multi-line text input with a hint icon | Use this type for [non-Boolean expressions](https://surveyjs.io/Documentation/Library?id=LibraryOverview#conditions-functions). |
| `"number"` | Text input | Use this type for numeric values. |
| `"text"` | Multi-line text input | Use this type for multi-line text values. |
| `"html"` | Multi-line text input | Use this type for HTML markup. |
| `"itemvalues"` | Customized text inputs for entering value-text pairs | Use this type for arrays of objects with the following structure: `{ value: any, text: string }`. For example, Dropdown, Checkbox, and Radiogroup questions use this type for the [`choices`](https://surveyjs.io/Documentation/Library?id=QuestionSelectBase#choices) property. |
| `"value"` | Button that opens a dialog window | The dialog window displays the survey element and allows users to set the element's default value. |

You can add the type to the `name` property after a colon character as a shortcut:

```js
Survey.Serializer.addProperty("question", 
  { name: "my-boolean-property", type: "boolean" }
  // ===== or =====
  { name: "my-boolean-property:boolean" }
);
```

#### `default`

> A default value for the property. If not specified, `default` equals to `""` for string values, 0 for numbers, `false` for Boolean values. The default value is not serialized into survey JSON definition.

Reviewer: ```suggestion
A default value for the property. If not specified, `default` equals `""` for string values, 0 for numbers, and `false` for Boolean values. The default value is not serialized into a survey JSON definition.
```

An array of selection choices or a function that loads the choices from a web service. Applies only to text and numeric properties. If `choices` are specified, Survey Creator renders a drop-down menu as the property editor.

```js
// Define `choices` locally
Survey.Serializer.addProperty("question", {
  name: "my-string-property",
  choices: [ "option1", "option2", "option3" ],
  // If item captions should be different from item values:
  // choices: [
  //   { value: "option1", text: "Option 1" },
  //   { value: "option2", text: "Option 2" },
  //   { value: "option3", text: "Option 3" },
  // ],
  default: "option1"
});

// Load `choices` from a web service
Survey.Serializer.addProperty("survey", {
  name: "country",
  category: "general",
  choices: function (obj, choicesCallback) {
    const xhr = new XMLHttpRequest();
    xhr.open("GET", "https://surveyjs.io/api/CountriesExample");
    xhr.setRequestHeader( "Content-Type", "application/x-www-form-urlencoded");
    xhr.onload = function () {
      if (xhr.status === 200) {
        const response = JSON.parse(xhr.response);
        const result = [];
        // Make the property nullable
        result.push({ value: null });

>         // Web service returns objects that we convert to the `{ value, text }` format

Reviewer: Is it necessary to say "we"? For example, can we say the following?

`// Web service returns objects that are converted to the `{ value, text }` format`

>         // If your web service returns an array of strings, simply pass it to `choicesCallback`

Reviewer: 1. Please remove the word "simply".

2. Does "it" refer to the "web service"?

  // ===== or =====
  { name: "!my-boolean-property", type: "boolean" }
);
```

#### `isSerializable`

A Boolean value that specifies whether to include the property in the survey JSON definition. Defaults to `true`.

#### `isLocalizable`

A Boolean value that specifies whether users can translate the property value to different languages in the Translation tab. Applies only to text properties. Defaults to `false`.

```js
Survey.Serializer.addProperty("question", 
  { name: "my-text-property", type: "text", isLocalizable: true }
);
```

#### `visible`

A Boolean value that specifies whether the property is visible in the Property Grid. Defaults to `true`.

#### `visibleIf`

A function that specifies a condition based on which to show or hide the property. The function accepts the question or panel that a user configures as a parameter.

If the property visibility depends on another property, use the [`dependsOn`](#dependson) setting. You can use `visibleIf` in conjunction with `dependsOn` to impose more specific rules on property visibility. In this case, Survey Creator calls the `visibleIf` function only when one of the properties from the `dependsOn` array is changed.

> In the following code, the `dateFormat` property depends on the `inputType` property and is visible only if `inputType` is set to one of date types:

Reviewer: ```suggestion
In the following code, the `dateFormat` property depends on the `inputType` property and is visible only if `inputType` is set to one of the date types:
```

```js
Survey.Serializer.addProperty("survey", {
  name: "region",
  category: "Region",
  categoryIndex: 1,
  choices: ["Africa", "Americas", "Asia", "Europe", "Oceania"],
});

Survey.Serializer.addProperty("survey", {
  name: "country",
  category: "Region",
  dependsOn: [ "region" ],
  choices: function (obj, choicesCallback) {
    const xhr = new XMLHttpRequest();
    const url =
      !!obj && !!obj.region
        ? "https://surveyjs.io/api/CountriesExample?region=" + obj.region
        : "https://surveyjs.io/api/CountriesExample";
    xhr.open("GET", url);
    xhr.setRequestHeader("Content-Type","application/x-www-form-urlencoded");
    xhr.onload = function () {
      if (xhr.status === 200) {
        const response = JSON.parse(xhr.response);
        const result = [];
        // Make the property nullable
        res.push({ value: null });
        response.forEach(item => {
          result.push({ value: item.cioc, text: item.name });
        });
        choicesCallback(result);
      }
    };
    xhr.send();
  }
});
```

#### `onGetValue`

A function that you can use to adjust or exclude the property value from the survey JSON definition.

```js

> // get title property returns a title with question number and so on "5) My super title.",

Reviewer: Do you mean the following?

// The get title property returns a title with its question number (for example, "5) My super title."),

> //but we want to serialize only a "pure" question title "My super title".

Reviewer: Is it necessary to say "pure"? For example, can we say the following?

`// but you want to serialize only the question title ("My super title.").`

{ name: "title:text", onGetValue: function (obj: any) { return obj.titleValue; } }

> // the function always returns null. It means that the library will never serialize the property in JSON.

Reviewer: ```suggestion
// The function always returns null. This means that the library will never serialize the property in JSON.
```

{ name: "calcProperty", onGetValue: function (obj: any) { return null; } }
```
#### `onSetValue`

> A function that you can use to perform actions when the property value is set, for example, update another property value.

Reviewer: ```suggestion
A function that you can use to perform actions when the property value is set (for example, update another property value).
```

> NOTE: Do not assign a value directly to an object property because this will trigger the `onSetValue` function again. Use the object's `setPropertyValue(propertyName, newValue)` method instead.

```js
{ 
    name: "myValue", 
    onSetValue: function (obj, value) {

>         //Optionally do some checks, modify value if needed.

Reviewer: ```suggestion
        //Optionally perform some checks and modify the value if needed.
```

        //Set value directly
        obj.setPropertyValue("myValue", value);

>         //Perform some actions, for example change other properties of the object

Reviewer: ```suggestion
        //Perform some actions (for example change other properties of the object).
```
